controllers/servicefence_controller.go

```diff
@@ -313,26 +369,135 @@ func (r *ServicefenceReconciler) updateVisitedHostStatus(host *lazyloadv1alpha1.
 		}
 		return lazyloadv1alpha1.Destinations_ACTIVE
 	}
-
-	domains := make(map[string]*lazyloadv1alpha1.Destinations)
 	now := time.Now().Unix()
 
-	// update status of destinations of hosts with config
-	for h, strategy := range host.Spec.Host {
-		allHost := []string{h}
-		if hs := getDestination(h); len(hs) > 0 {
+	for h, strategy := range sf.Spec.Host {
+
+		fullHost := h
+		subDomains := strings.Split(h, ".")
+		switch len(subDomains) {
+		case 5:
+		case 1:
+			fullHost = fmt.Sprintf("%s.%s.svc.cluster.local", subDomains[0], sf.Namespace)
+		case 2:
+			fullHost = fmt.Sprintf("%s.%s.svc.cluster.local", subDomains[0], subDomains[1])
+		default:
+			log.Errorf("%s is invalid host, skip", h)
+			continue
+		}
+		if !isValidHost(fullHost) {
+			continue
+		}
+		if domains[fullHost] != nil {
+			continue
+		}
+
+		allHost := []string{fullHost}
+		if hs := getDestination(fullHost); len(hs) > 0 {
 			allHost = append(allHost, hs...)
 		}
 
-		domains[h] = &lazyloadv1alpha1.Destinations{
+		domains[fullHost] = &lazyloadv1alpha1.Destinations{
 			Hosts:  allHost,
 			Status: checkStatus(now, strategy),
 		}
 	}
+}
 
-	// update status with metric status
-	// -> (visited)host -> destinations
-	for metricName := range host.Status.MetricStatus {
+// update domains with spec.namespaceSelector
+func genDomainsWithNamespaceSelector(domains map[string]*lazyloadv1alpha1.Destinations, sf *lazyloadv1alpha1.ServiceFence,
+	clientSet *kubernetes.Clientset) {
+	for _, ns := range sf.Spec.NamespaceSelector {
+		svcs, err := clientSet.CoreV1().Services(ns).List(metav1.ListOptions{})
+		if err != nil {
+			log.Errorf("failed to list services of namespace %s, skip", ns)
+			continue
+		}
+		for _, svc := range svcs.Items {
+			fullHost := fmt.Sprintf("%s.%s.svc.cluster.local", svc.GetName(), svc.GetNamespace())
+			if !isValidHost(fullHost) {
+				continue
+			}
+			if domains[fullHost] != nil {
+				continue
+			}
+			allHost := []string{fullHost}
+			if hs := getDestination(fullHost); len(hs) > 0 {
+				allHost = append(allHost, hs...)
+			}
+			domains[fullHost] = &lazyloadv1alpha1.Destinations{
+				Hosts:  allHost,
+				Status: lazyloadv1alpha1.Destinations_ACTIVE,
+			}
+		}
+	}
+}
+
+// update domains with spec.labelSelector
+func genDomainsWithLabelSelector(domains map[string]*lazyloadv1alpha1.Destinations, sf *lazyloadv1alpha1.ServiceFence,
+	labelSvcCache map[string][]string, svcCacheLock *sync.RWMutex) {
+
+	svcCacheLock.RLock()
+	defer svcCacheLock.RUnlock()
+
+	for _, selector := range sf.Spec.LabelSelector {
+
+		// get services matched all labels of this selector
+		var tmpSvcsSlice [][]string
+		var result []string
```

Comment:
a `map[string]struct{}` is enough to collect and de-dup

Reply:
use `result := make(map[string]struct{})`  to check and get result for domains